main/src/library/Chain.flix

```diff
@@ -0,0 +1,847 @@
+    ///
+    pub enum ViewRight[t] with Eq {
+        case NoneRight
+        case SomeRight(Chain[t], t)
+    }
+
+    ///
+    /// Return the empty chain.
+    ///
+    pub def empty(): Chain[a] = Empty
+
+    ///
+    /// Return the singleton chain with element `x`.
+    ///
+    pub def point(x: a): Chain[a] = One(x)
+
+    ///
+    /// Apply every function from `f` to every argument from `x` and return a chain with all results.
+    /// For `f = f1, f2, ...` and `x = x1, x2, ...` the results appear in the order
+    /// `f1(x1), f1(x2), ..., f2(x1), f2(x2), ...`.
+    ///
+    pub def ap(f: Chain[a -> b & ef], x: Chain[a]) : Chain[b] & ef =
+        apHelper(viewLeft(f), x, empty())
+
+    ///
+    /// Helper function for `ap`.
+    ///
+    def apHelper(f: ViewLeft[a -> b & ef], x: Chain[a], ac: Chain[b]): Chain[b] & ef = match f {
+        case NoneLeft         => ac
+        case SomeLeft(f1, rs) => apHelper(viewLeft(rs), x, ac `append` map(f1, x))
+    }
+
+    ///
+    /// Returns true if and only if `xs` is the empty chain, i.e. `Empty`.
+    ///
+    pub def isEmpty(xs: Chain[a]): Bool = match xs {
+        case Empty => true
+        case _     => false
+    }
+
+    ///
+    /// Add element `x` to the left end of Chain `xs`.
+    ///
+    pub def cons(x: a, xs: Chain[a]): Chain[a] = match xs {
+        case Empty => One(x)
+        case _     => Chain(One(x), xs)
+    }
+
+    ///
+    /// Add element `x` to the right end of Chain `xs`.
+    ///
+    pub def snoc(xs: Chain[a], x: a): Chain[a] = match xs {
+        case Empty => One(x)
+        case _     => Chain(xs, One(x))
+    }
+
+    ///
+    /// Returns `Some(x)` if `x` is the first element of `xs`.
+    ///
+    /// Returns `None` if `xs` is empty.
+    ///
+    pub def head(xs: Chain[a]): Option[a] = match viewLeft(xs) {
+        case SomeLeft(x, _) => Some(x)
+        case _              => None
+    }
+
+    ///
+    /// Returns `Some(x)` if `x` is the last element of `xs`.
+    ///
+    /// Returns `None` if `xs` is empty.
+    ///
+    pub def last(xs: Chain[a]): Option[a] = match viewRight(xs) {
+        case SomeRight(_, x) => Some(x)
+        case _               => None
+    }
+
+    ///
+    /// Returns the sublist of `xs` without the last element.
+    /// Returns `None` if the chain `xs` is empty.
+    ///
+    pub def init(xs: Chain[a]): Option[Chain[a]] = match viewRight(xs) {
+        case SomeRight(rs, _) => Some(rs)
+        case _                => None
+    }
+
+    ///
+    /// Returns the length of `xs`.
+    ///
+    pub def length(xs: Chain[a]): Int32 = foldRight((_, ac) -> ac + 1, 0, xs)
+
+    ///
+    /// Returns `ys` appended to `xs`.
+    ///
+    pub def append(xs: Chain[a], ys: Chain[a]): Chain[a] = match (xs, ys) {
+        case (Empty, _) => ys
+        case (_, Empty) => xs
+        case (_, _)     => Chain(xs, ys)
+    }
+
+    ///
+    /// Deconstruct a Chain from the left end.
```

Comment:
Could you add a bit more intuition. I assume this means that we are seeing elements from left to right. (And the same for viewRight)?

Reply:
Fixed for `viewLeft` and `viewRight`